# Recursion

It is a technique, where a function calls itself.

Benefits:
- No need for loop
- Mostly asked in interviews

Disadvantages
- Difficult thing
- Not used in practical cases much in day to day tasks.

In [ ]:
# multiplying using loop
def multiply(a, b):
  result = 0

  for i in range(b):
    result += a
  print(result)

multiply(4, 5)

20


Recursion requires two things: 1. Base Case; and, 2. Decomposition of a big problem into smaller problems. Now, writing the above code using **Recursion**.

In [9]:
def mul(a, b):
  if b == 1:
    return a
  else:
    return a + (mul(a, b-1))
mul(4,3)

12

The `mul(a, b)` function calculates the product of `a` and `b` using recursion. In this version of multiplication, instead of directly multiplying `a` and `b`, the function adds `a` repeatedly to the result of `mul(a, b-1)` until `b` reaches 1, which is the base case.

### Breakdown of recursive calls for `mul(4, 3)`:

1. `mul(4, 3)` calls `mul(4, 2)`
2. `mul(4, 2)` calls `mul(4, 1)`
3. `mul(4, 1)` hits the base case and returns `4`

Now, as the recursion unwinds:

- `mul(4, 1)` returns 4
- `mul(4, 2)` returns `4 + mul(4, 1)` = `4 + 4` = 8
- `mul(4, 3)` returns `4 + mul(4, 2)` = `4 + 8` = 12

So, the result of `mul(4, 3)` is `12`.

### Recursion Tree for `mul(4, 3)`:

```
                       mul(4, 3)
                         /    
                    mul(4, 2)
                    /    
                mul(4, 1)   <-- returns 4
```

### How the recursion "unwinds":
1. `mul(4, 1)` returns `4`
2. `mul(4, 2)` returns `4 + 4` = `8`
3. `mul(4, 3)` returns `4 + 8` = `12`

The final result of `mul(4, 3)` is `12`. This process adds `a` repeatedly `b` times.

Here, in the above code, the concept being used is known as **unwinding** or **recursive return**.
A "**recursive return**" or "**returning the value during the unwinding of the recursion**". This happens when a function calls itself recursively, and once the base case is reached, the recursion starts returning values as the call stack unwinds.

In [ ]:
# factorial using recursion
def fact(num):
  if num == 1:
    return 1
  else:
    return num * fact(num-1)
fact(5)

120

The function `fact(num)` calculates the factorial of `num`. When `fact(5)` is called, the recursion proceeds as follows:

1. `fact(5)` calls `fact(4)`
2. `fact(4)` calls `fact(3)`
3. `fact(3)` calls `fact(2)`
4. `fact(2)` calls `fact(1)`
5. `fact(1)` reaches the base case and returns 1.

Then the recursive calls start returning, and the values are multiplied as the recursion tree "unwinds":

- `fact(1)` returns 1
- `fact(2)` returns `2 * fact(1)` = `2 * 1` = 2
- `fact(3)` returns `3 * fact(2)` = `3 * 2` = 6
- `fact(4)` returns `4 * fact(3)` = `4 * 6` = 24
- `fact(5)` returns `5 * fact(4)` = `5 * 24` = 120

### Recursion Tree for `fact(5)`:

```
                              fact(5)
                             /    
                         fact(4)
                        /    
                    fact(3)
                   /    
               fact(2)
              /    
          fact(1)  (returns 1)
```

Now, the recursion unwinds:

```
fact(1) = 1
fact(2) = 2 * 1 = 2
fact(3) = 3 * 2 = 6
fact(4) = 4 * 6 = 24
fact(5) = 5 * 24 = 120
```

This tree shows how the recursion works: each call waits for the result of the next recursive call to multiply by the current `num`. When the base case is hit (fact(1)), it begins to resolve each call, and the values get multiplied on the way back up the recursion tree.

In [11]:
def is_palindrome(s):
    # Base case: if the string is empty or a single character, it's a palindrome
    if len(s) <= 1:
        return True
    # Check the first and last character
    elif s[0] == s[-1]:
        # Recursively check the substring without the first and last character
        return is_palindrome(s[1:-1])
    else:
        return False

# Example call
is_palindrome("racecar")  # Should return True

True

A **palindrome** is a word, phrase, or sequence that reads the same backward as forward (ignoring spaces, punctuation, and capitalization). For this program, we'll focus on checking whether a given string is the same forwards and backwards.

### Recursive breakdown for `is_palindrome("racecar")`:

1. `is_palindrome("racecar")` compares `'r'` with `'r'`. Since they are equal, it calls `is_palindrome("aceca")`.
2. `is_palindrome("aceca")` compares `'a'` with `'a'`. Since they are equal, it calls `is_palindrome("cec")`.
3. `is_palindrome("cec")` compares `'c'` with `'c'`. Since they are equal, it calls `is_palindrome("e")`.
4. `is_palindrome("e")` reaches the base case (`len(s) <= 1`) and returns `True`.

Now, as each recursive call returns:

- `is_palindrome("e")` returns `True`
- `is_palindrome("cec")` returns `True` (because both the first and last characters matched, so it continues checking the inner string)
- `is_palindrome("aceca")` returns `True`
- `is_palindrome("racecar")` returns `True`

Thus, `is_palindrome("racecar")` returns `True`, meaning the string is a palindrome.

### Recursion Tree for `is_palindrome("racecar")`:

```
                       is_palindrome("racecar")
                             /        \
               s[0] == s[-1] -> True
                             /           \
                       is_palindrome("aceca")
                             /        \
               s[0] == s[-1] -> True
                             /           \
                       is_palindrome("cec")
                             /        \
               s[0] == s[-1] -> True
                             /           \
                       is_palindrome("e")    <-- returns True (base case)
```

### How the recursion works:
1. The first and last characters of the string are checked.
2. If they match, the function is called recursively on the substring excluding the first and last characters.
3. If at any point the characters don’t match, the function returns `False`.
4. The recursion continues until the base case is reached (`len(s) <= 1`), where it returns `True` indicating a palindrome.

### Example of Non-Palindrome (`is_palindrome("hello")`):

For a string like `"hello"`, the recursion works similarly, but at some point, the characters will not match:

1. `is_palindrome("hello")` compares `'h'` with `'o'` and finds they do not match. Therefore, it immediately returns `False`.

Let me know if you need further clarification!

### Space-Time Trade-off with Fibonacci Sequence:

The **space-time trade-off** is a concept in computing where you optimize the time complexity (how fast the program runs) by using additional space (memory). This is often done by storing intermediate results in memory to avoid recalculating them, which can significantly speed up computations at the cost of using more memory.

When calculating the Fibonacci sequence for **large numbers**, a **naive recursive approach** becomes extremely inefficient due to the number of redundant function calls. Let's break this down:

### Why the usual Fibonacci code is inefficient for large numbers:

The **naive Fibonacci** code typically uses recursion to calculate Fibonacci numbers like this:

```python
def fibonacci(n):
    if n <= 1:
        return n
    else:
        return fibonacci(n-1) + fibonacci(n-2)
```

#### **Problem with naive recursion:**
1. **Exponential Time Complexity:**
   - This naive approach has an **exponential time complexity** of \( O(2^n) \) because each function call results in two more calls.
   - For example, to compute `fibonacci(5)`, you compute `fibonacci(4)` and `fibonacci(3)`, which recursively compute their own results.
   - Many of these recursive calls calculate the same Fibonacci numbers multiple times. For instance, `fibonacci(3)` is computed twice, `fibonacci(2)` is computed multiple times, and so on.
   - This redundancy leads to **inefficiency**, and the program becomes slow as `n` increases. For larger values of `n`, this approach becomes impractical due to the huge number of recursive calls.

#### **Example:**

If you try to calculate `fibonacci(50)` using the naive approach, it will involve millions of redundant calculations, resulting in an extremely slow execution time.

### **Memoization to the Rescue:**

**Memoization** is a technique used to optimize recursive algorithms by storing the results of function calls in a cache (usually a dictionary or array). When a function is called again with the same parameters, instead of recalculating the result, it simply returns the stored value from the cache.

#### **How Memoization Works:**

- **Before** a recursive call, we check if the result for a given input (`n`) is already computed and stored in a cache (a dictionary, for instance).
- **If the result is found** in the cache, we simply return it instead of doing the calculation again.
- **If not**, we perform the calculation, store the result in the cache, and return the result.

This approach dramatically reduces the time complexity by avoiding the repeated recalculations.

### **How Memoization Solves the Fibonacci Problem:**

Consider the memoized Fibonacci code you provided:

```python
import time
def memo(m, d):
    if m in d:  # Check if the result is already computed and stored in the dictionary
        return d[m]
    else:
        d[m] = memo(m-1, d) + memo(m-2, d)  # Recursively compute and store the result
        return d[m]

start = time.time()
d = {0: 1, 1: 1}  # Initialize the memoization dictionary with base cases
print(memo(12, d))  # Compute the 12th Fibonacci number
print(time.time() - start)  # Print the time taken for computation
```

### **Why the Given Code is Efficient:**

1. **Caching Results:**
   - The dictionary `d` stores previously computed Fibonacci numbers.
   - For example, when calculating `memo(12)`, the function will compute `memo(11)` and `memo(10)`. However, if `memo(10)` has already been computed (from the call to `memo(11)`), the function will fetch the result from the cache instead of recalculating it.

2. **Eliminating Redundant Computations:**
   - Without memoization, for each Fibonacci number `n`, the program would recursively call `fibonacci(n-1)` and `fibonacci(n-2)`, which could lead to recalculating the same Fibonacci numbers multiple times.
   - With memoization, once a number is computed (like `fibonacci(10)`), it is stored in `d`. In subsequent calls, the program simply returns the value from the dictionary, thus reducing the number of calls exponentially.

3. **Time Complexity with Memoization:**
   - The time complexity of this optimized approach is \( O(n) \), where \( n \) is the Fibonacci number you are trying to calculate.
   - Every Fibonacci number from 0 to `m` is computed **once**, and subsequent lookups are **constant time** \( O(1) \) due to dictionary access.
   - This is a **significant improvement** over the naive recursive approach, which had exponential time complexity.

4. **Space Complexity:**
   - Memoization requires additional space to store the results of previously computed Fibonacci numbers. The space complexity is \( O(n) \) due to the dictionary `d`.

5. **Performance:**
   - As you can see from the code, using memoization makes the fetching of Fibonacci numbers **extremely fast**. The time taken to compute the Fibonacci number drops significantly because the algorithm avoids redundant recalculations. In the example, for `memo(12)`, the time complexity is linear \( O(n) \), and the execution time will be quite small.

### **Example of Memoization Performance:**

#### **Without Memoization:**
Calculating `fibonacci(12)` would involve many redundant calls and would take longer as the number grows. The time complexity grows exponentially with the input size.

#### **With Memoization:**
By storing results in a dictionary, the time complexity reduces to \( O(n) \), and the program quickly computes the result without redundant calls.

### **Why is Memoization Fast and Efficient?**

- **Avoiding Redundant Work:**
  - The main reason why memoization makes the Fibonacci calculation fast is that it prevents **recalculating the same values** over and over again.
  
- **Dictionary Lookup:**
  - Accessing a dictionary in Python is a constant-time operation, i.e., \( O(1) \). When checking if a Fibonacci number has already been computed, it takes constant time.

- **Efficient Recursion:**
  - Instead of recalculating values for every recursive call, the results are cached and reused, making the function calls much more efficient.

### Conclusion:

- **Memoization** is a powerful technique that enhances the efficiency of recursive algorithms by storing intermediate results.
- It turns an **exponentially time-consuming** Fibonacci function into a **linear-time** function.
- By trading space (using memory to store intermediate results), we achieve a **significant reduction in time complexity**, making the algorithm feasible even for large inputs.

This technique is widely applicable for problems where overlapping subproblems occur, and it makes algorithms **extremely efficient** and faster by optimizing the use of memory.

In [15]:
import time
def memo(m, d):

  if m in d:
    return d[m]
  else:
    d[m] = memo(m-1, d) + memo(m-2, d)
    return d[m]

start = time.time()
d = {0:1, 1:1}
print(memo(12, d))
print(time.time()-start)

233
0.00021076202392578125


Suppose we do this same thing, but without memoization, and find the fibonacci sequence up till `12`, using a function `fib()`. 

- In such a case, the program will first find `fib(11)` to `fib(1)`.
- In each of those `fib()`, the program will calculate `fib()` for each number, back till `1`.

Why is dynamic programming used? Make a powerset using recursion.